### PR TITLE
Languages on a item for client with uls

### PR DESCRIPTION
**Pull request by bitPogo**

### CODE REVIEW

File: src/client/data-access/UlsLanguageRepository.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	private contentLanguages: WikibaseContentLanguages;
+	private directionalities: UlsData;
+
+	constructor( contentLanguages: WikibaseContentLanguages, directionalities: UlsData ) {
+		this.contentLanguages = contentLanguages;
+		this.directionalities = directionalities;
+	}
+
+	public getLanguages(): Promise<LanguageCollection> {
+		return new Promise<LanguageCollection>( ( resolve ) => {
```
Comment (jakobw) [inline src/client/data-access/UlsLanguageRepository.ts]:
```suggestion
		return Promise.resolve( this.getLanguageCollection() );
```

File: src/client/data-access/UlsLanguageRepository.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	private getLanguageCollection(): LanguageCollection {
+		const codes = this.getLanguagesCodes();
+		const collection = {} as LanguageCollection;
+		codes.forEach( ( code ) => {
+			collection[ code ] = this.getDircectionalityByKey( code );
+		} );
+		return collection;
+	}
+
+	private getLanguagesCodes(): string[] {
```
Comment (jakobw) [inline src/client/data-access/UlsLanguageRepository.ts]:
```suggestion
	private getLanguageCodes(): string[] {
```

File: tests/unit/client/data-access/UlsLanguageRepository.spec.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+		};
+		const directionalities = {
+			getDir: ( code: string ) => langDirs[ code ],
+		};
+		return (
+			new UlsLanguageRepository(
+				translator,
+				directionalities,
+		) ).getLanguages().then( ( name ) => {
+			expect( name ).toEqual( {
```
Comment (wiese) [inline tests/unit/client/data-access/UlsLanguageRepository.spec.ts]:
```suggestion
			expect( languages ).toEqual( {
```

File: src/client/data-access/UlsLanguageRepository.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+	public getLanguages(): Promise<LanguageCollection> {
+		return new Promise<LanguageCollection>( ( resolve ) => {
+			resolve( this.getLanguageCollection() );
+		} );
+	}
+
+	private getLanguageCollection(): LanguageCollection {
+		const codes = this.getLanguagesCodes();
+		const collection = {} as LanguageCollection;
```
Comment (jakobw) [inline src/client/data-access/UlsLanguageRepository.ts]:
```suggestion
		const collection: LanguageCollection = {};
```

File: src/client-entry.ts [hunk 5]
```diff
@@ -4,16 +4,26 @@ import buildApp from '@/common/buildApp';
...
-	new UlsLanguageTranslationRepository( new ( window as MwWindow ).wb.WikibaseContentLanguages() ),
+	new UlsLanguageTranslationRepository(
+		new ( window as MwWindow ).wb.WikibaseContentLanguages(),
+	),
 );
-// TODO factory.setLanguageRepository( languageRepo );
+
+factory.setLanguageRepository(
+	new UlsLanguageRepository(
+		new ( window as MwWindow ).wb.WikibaseContentLanguages(),
```
Comment (wiese) [inline src/client-entry.ts]: If we consider `WikibaseContentLanguages` a service (I do), we should reuse it.

File: src/client/data-access/UlsLanguageRepository.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+		return collection;
+	}
+
+	private getLanguagesCodes(): string[] {
+		return Object.keys(
+			this.contentLanguages.getAllPairs(),
+		);
+	}
+
+	private getDircectionalityByKey( languageCode: string ): Language {
```
Comment (jakobw) [inline src/client/data-access/UlsLanguageRepository.ts]:
```suggestion
	private getDirectionalityByKey( languageCode: string ): Language {
		return {
			code: languageCode,
			directionality: this.directionalities.getDir( languageCode ),
		};
	}
```

File: src/client/data-access/UlsLanguageRepository.ts (hunk 1 above)
Comment (wiese) [inline src/client/data-access/UlsLanguageRepository.ts]:
```suggestion
		return Promise.resolve( this.getLanguageCollection() );
```

File: src/mockup-entry.ts [hunk 7]
```diff
@@ -19,3 +20,11 @@ class MockupWikibaseContentLanguages {
 ( window as MwWindow ).wb = {
 	WikibaseContentLanguages: MockupWikibaseContentLanguages,
 };
+
+( window as MwWindow ).$ = {
+	uls: {
+		data: {
+			getDir: ( code: string ) => directionalities.default[ code ],
```
Comment (jakobw) [inline src/mockup-entry.ts]: I would have written a fake implementation that would not have required the additional json file, but I think this is ok too.

File: tests/unit/client/data-access/UlsLanguageRepository.spec.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
+import UlsLanguageRepository from '@/client/data-access/UlsLanguageRepository';
+import { StringTMap } from '@/datamodel/LanguageTranslations';
+
+describe( 'UlsLanguageRepository', () => {
+
+	it( 'gets the language name from the given function', () => {
```
Comment (jakobw) [inline tests/unit/client/data-access/UlsLanguageRepository.spec.ts]: confusing test description. should mention Wikibase and ULS maybe, not "given function"

File: src/client-entry.ts (hunk 5 above)
Comment (bitPogo) [reply src/client-entry.ts]: agree

**Pull request merged**